src/main/java/org/spongepowered/api/command/parameter/Parameter.java

```diff
@@ -0,0 +1,1236 @@
+
+import javax.annotation.Nullable;
+
+/**
+ * Defines how an element of a command argument string should be parsed.
+ *
+ * <p>Parameters that parse input and return an object into the
+ * {@link CommandContext} should be of type {@link Parameter.Value} in
+ * order to maximize type safety. However, there are also other important
+ * ways that parameters can be used:</p>
+ *
+ * <ul>
+ *     <li>{@link #firstOf(Parameter)} allows for multiple parameters that
+ *     do not have the same return type to attempt to parse an input
+ *     successfully.</li>
+ *     <li>{@link #seq(Parameter)} allows for the grouping of multiple
+ *     parameters that will be executed one after another.</li>
+ *     <li>{@link Subcommand}s can be placed anywhere in a parameter
+ *     chain where a {@link Parameter} can be added, if successfully parsed,
+ *     any containing {@link Command} would take precedence and its
+ *     {@link Command#process(Cause, String)} method will be called instead
+ *     of any parent.</li>
+ * </ul>
+ *
+ * <p>{@link Parameter}s are intended for use with {@link Command.Builder}s.
+ * </p>
+ */
+public interface Parameter {
+
+    /**
+     * Creates a {@link Parameter.Key} for storing values against.
+     *
+     * @param key The string key
+     * @param valueClass The type of value that this key represents
+     * @param <T> The type
+     * @return The {@link Key}
+     */
+    static <T> Key<T> key(String key, Class<T> valueClass) {
+        return Sponge.getRegistry().createBuilder(Key.Builder.class).build(key, valueClass);
+    }
+
+    /**
+     * Gets a builder that builds a {@link Parameter.Value}.
+     *
+     * @return The {@link Value.Builder}
+     */
+    static <T> Value.Builder<T> builder(Class<T> valueClass) {
+        return ((Value.Builder<T>) Sponge.getRegistry().createBuilder(Value.Builder.class)).setValueClass(valueClass);
+    }
+
+    /**
+     * Gets a builder that builds a {@link Parameter.Value}.
+     *
+     * @return The {@link Value.Builder}
+     */
+    static <T> Value.Builder<T> builder(Class<T> valueClass, ValueParameter<T> parameter) {
+        return ((Value.Builder<T>) Sponge.getRegistry().createBuilder(Value.Builder.class)).setValueClass(valueClass).parser(parameter);
+    }
+
+    /**
+     * Gets a {@link Parameter} that represents a subcommand.
+     *
+     * <p>If a {@link Subcommand} alias in a parameter chain
+     * is successfully matched, then element parsing will continue
+     * with the parameters supplied to {@link Command}. It is
+     * implementation dependent as to what happens if a subcommand
+     * fails to parse.</p>
+     *
+     * @param subcommand The {@link Command} to execute
+     * @param alias The first alias of the subcommand
+     * @param aliases Subsequent aliases, if any
+     * @return The {@link Subcommand} for use in a {@link Parameter} chain
+     */
+    static Subcommand subcommand(Command subcommand, String alias, String... aliases) {
```

Comment:
alias -> primaryAlias
aliases -> secondaryAliases

Reply:
No guarantee that `alias` will become the primary alias.

Reply:
Hmm... For sure! This is a subcommand.

Reply:
Okay, I misread this - but I'm really not sold on that naming - primary/secondary alias is not appropriate here. It's just a set of aliases, this is a way to guarantee one alias and thus flag up to IDEs if the alias is forgotten. That's all.

alias/aliases is fine here. Javadocs should describe this enough.